# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Due Date:** [Insert Due Date Here]

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [56]:
# Your code here:
# 1. Load the tidyverse package
#Load tidyverse
# # Load necessary packages for data manipulation and joining
library(tidyverse) 

# 2. Set working directory to sample_datasets 
#getwd() 
setwd("C:/Users/franc/OneDrive/Desktop/Documents/MS 3083/Data")

# 3. Import all CSV files using read_csv() with EXACT variable names above
customers <- read.csv("customers.csv")
orders <- read.csv("orders.csv")
order_items <- read.csv("order_items.csv")
products <- read.csv("products.csv")
suppliers <- read.csv("suppliers.csv")

# Print dataset information (uncomment after creating variables above):
cat("Dataset Dimensions:\n")
cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

#Show first 3 rows of each dataset (uncomment after loading data):
head(customers, 3)
head(orders, 3)
head(order_items, 3)
head(products, 3)
head(suppliers, 3)


Dataset Dimensions:
Customers: 100 rows x 5 columns
Orders: 250 rows x 4 columns
Order Items: 400 rows x 4 columns
Products: 50 rows x 4 columns
Suppliers: 10 rows x 3 columns


CustomerID Name       Email               City        Registration_Date
1 1          Customer 1 customer1@email.com Phoenix     2020-10-03       
2 2          Customer 2 customer2@email.com Los Angeles 2020-06-02       
3 3          Customer 3 customer3@email.com Chicago     2021-04-20

OrderID CustomerID Order_Date Total_Amount
1 1       87         2023-08-30 424.30      
2 2       12         2024-03-24 183.09      
3 3       37         2024-03-19 549.07

OrderID ProductID Quantity Unit_Price
1 213      8        1         43.81    
2 176     18        5        489.16    
3 118      2        5        442.09

ProductID Product_Name Category    Supplier_ID
1 1         Product 1    Home        6          
2 2         Product 2    Electronics 7          
3 3         Product 3    Home        1

Supplier_ID Supplier_Name Country
1 1           Supplier 1    USA    
2 2           Supplier 2    USA    
3 3           Supplier 3    Japan

## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [57]:
cat("Customers columns:\n")
print(colnames(customers))
cat("\n")

cat("Orders columns:\n")
print(colnames(orders))
cat("\n")

Customers columns:
[1] "CustomerID"        "Name"              "Email"            
[4] "City"              "Registration_Date"

Orders columns:
[1] "OrderID"      "CustomerID"   "Order_Date"   "Total_Amount"



In [58]:
# Your code here:
# Create an inner join between customers and orders
customer_orders_inner <- inner_join(customers, orders, by ="CustomerID")
# Analyze the results and compare row counts
cat("  Customers:", nrow(customers), "rows\n")
cat("  Orders:", nrow(orders), "rows\n\n")


  Customers: 100 rows
  Orders: 250 rows



In [59]:
# Your code here:
# Create a left join to keep all customers
customer_orders_left <- left_join(customers, orders, by = "CustomerID")
# Count how many customers have no orders
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Customers without orders:", customers_without_orders, "\n\n")


Customers without orders: 0 



In [60]:
# Your code here:
# Create a right join to keep all orders
customer_orders_right <- right_join(customers, orders, by = "CustomerID")
# Check for orders with invalid customer IDs
orders_without_customers <- sum(is.na(customer_orders_right$Name))
cat("Orders without customer match:", orders_without_customers, "\n\n")




Orders without customer match: 50 



<VSCode.Cell id="#VSC-0da36c1e" language="markdown">
## Part 2: Basic Joins

**Required Variable Names:**
You must create these EXACT variable names:

**Part 2.1:** Create `customer_orders` using inner_join()
**Part 2.2:** Create `customer_orders_left` using left_join()  
**Part 2.3:** Create `customer_orders_right` using right_join()
**Part 2.4:** Create `customer_orders_full` using full_join()

**Expected Outputs:**
For each join, print:
- Number of rows in the result using `nrow()`
- Comparison with original dataset sizes
- Analysis of unmatched records

In [63]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- dplyr::inner_join(customers, orders, by = "CustomerID")


# Required output for autograding:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [64]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- dplyr::left_join(customers, orders, by = "CustomerID")

# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [65]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- dplyr::inner_join(customers, orders, by = "CustomerID")

# Expected output:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [66]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- dplyr::left_join(customers, orders, by = "CustomerID")

# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [70]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
# Your code here:
customer_orders_right <- dplyr::right_join(customers, orders, by = "CustomerID")

# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:
")
cat("Total rows:", nrow(customer_orders_right), "
")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "
")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [71]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- dplyr::full_join(customers, orders, by = "CustomerID")

# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:


Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [72]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- dplyr::full_join(customers, orders, by = "CustomerID")

# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:


Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [73]:
# Step 1: REQUIRED variable name: orders_items
# Your code here:
orders_items <- dplyr::left_join(order_items, orders, by = "OrderID")

# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [83]:
# Step 2: REQUIRED variable name: orders_customers_items  
# Your code here:
orders_customers_items <- dplyr::left_join(order_items, orders, by = "OrderID") %>%
  dplyr::left_join(customers, by = "CustomerID") %>%
  mutate(
    Quantity = suppressWarnings(as.numeric(Quantity)),
    Unit_Price = suppressWarnings(as.numeric(gsub("[$,]", "", as.character(Unit_Price)))),
    line_total = ifelse(is.na(Quantity) | is.na(Unit_Price), 0, Quantity * Unit_Price)
  )

# Required output for autograding:
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 400 rows


In [84]:
# Step 3: REQUIRED variable name: complete_order_data
# Your code here:
complete_order_data <- dplyr::left_join(
  orders_customers_items,
  products %>% select(ProductID, Product_Name, Category, Supplier_ID),
  by = "ProductID"
)

# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 400 rows


In [85]:
# Step 4: REQUIRED variable name: complete_data
# Your code here:
complete_data <- dplyr::left_join(
  complete_order_data,
  suppliers %>% select(Supplier_ID, Supplier_Name, Country),
  by = "Supplier_ID"
)

# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 400 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

<VSCode.Cell id="#VSC-d7f91846" language="markdown">
## Part 4: Data Quality Analysis

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customers_no_orders` - Use anti_join() to find customers with no orders
2. `orphaned_orders` - Use anti_join() to find orders without customers  
3. `products_never_ordered` - Use anti_join() to find unordered products
4. `active_customers` - Use semi_join() to find customers with orders

**Expected Outputs:**
Print the count of records for each using `nrow()`.

In [86]:
# REQUIRED variable name: customers_no_orders
# Your code here:
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [87]:
# REQUIRED variable name: orphaned_orders
# Your code here:
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")

# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [88]:
# REQUIRED variable name: products_never_ordered
# Your code here:
products_never_ordered <- anti_join(products, order_items, by = "ProductID")

# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [89]:
# REQUIRED variable name: active_customers
# Your code here:
active_customers <- semi_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [90]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset
# customers
dup_customers <- if ("CustomerID" %in% names(customers)) {
  nrow(customers) - n_distinct(customers$CustomerID)
} else { NA_integer_ }

# orders
dup_orders <- if ("OrderID" %in% names(orders)) {
  nrow(orders) - n_distinct(orders$OrderID)
} else { NA_integer_ }

# order_items (composite key OrderID + ProductID)
dup_order_items <- if (all(c("OrderID", "ProductID") %in% names(order_items))) {
  composite_keys <- paste0(as.character(order_items$OrderID), "||", as.character(order_items$ProductID))
  nrow(order_items) - n_distinct(composite_keys)
} else { NA_integer_ }

# products
dup_products <- if ("ProductID" %in% names(products)) {
  nrow(products) - n_distinct(products$ProductID)
} else { NA_integer_ }

# suppliers
dup_suppliers <- if ("Supplier_ID" %in% names(suppliers)) {
  nrow(suppliers) - n_distinct(suppliers$Supplier_ID)
} else { NA_integer_ }

# Required outputs for autograding (uncomment after creating the calculations above):
duplicate_customers <- sum(duplicated(customers$CustomerID))
duplicate_orders <- sum(duplicated(orders$OrderID))
duplicate_products <- sum(duplicated(products$ProductID))
duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

cat("Data Quality Summary:\n")
cat("Duplicate customer IDs:", duplicate_customers, "\n")
cat("Duplicate order IDs:", duplicate_orders, "\n")
cat("Duplicate product IDs:", duplicate_products, "\n")
cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

Data Quality Summary:
Duplicate customer IDs: 0 
Duplicate order IDs: 0 
Duplicate product IDs: 0 
Duplicate supplier IDs: 0 


## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

<VSCode.Cell id="#VSC-f7364f43" language="markdown">
## Part 5: Business Analysis with Joined Data

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customer_metrics` - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
2. `product_metrics` - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency  
3. `supplier_metrics` - Supplier analysis with Product_Count by supplier
4. `regional_analysis` - Sales analysis by customer city

**Required Columns:**
- customer_metrics: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
- product_metrics: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
- supplier_metrics: Supplier_ID, Supplier_Name, Product_Count
- regional_analysis: City, Total_Sales, Customer_Count, Avg_Customer_Value

In [ ]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
library(dplyr)

customer_metrics <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = dplyr::coalesce(Quantity, 0) * dplyr::coalesce(Unit_Price, 0)
  ) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  left_join(customers %>% select(CustomerID, Name), by = "CustomerID") %>%
  filter(!is.na(CustomerID)) %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(line_total, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  mutate(
    Avg_Order_Value = ifelse(Order_Count > 0, Total_Spent / Order_Count, 0)
  ) %>%
  select(CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value)

# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value


# Required output for autograding:
cat("Customer Analysis Summary:
")
cat("Total customers analyzed:", nrow(customer_metrics), "
")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "
")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

Customer Analysis Summary:
Total customers analyzed: 99 
Top customer total spent: $ 17221.69 
# A tibble: 3 × 5
  CustomerID Name  Total_Spent Order_Count Avg_Order_Value
       <int> <chr>       <dbl>       <int>           <dbl>
1        103 NA         17222.           9           1914.
2        102 NA         16116.           8           2015.
3        104 NA         13427.           7           1918.


In [ ]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
product_metrics <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(products %>% select(ProductID, Product_Name), by = "ProductID") %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(line_total, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  select(ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency)
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency


# Required output for autograding:
cat("Product Analysis Summary:
")
cat("Total products analyzed:", nrow(product_metrics), "
")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "
")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 13970.66 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <int> <chr>                <dbl>          <dbl>           <int>
1        26 Product 26          13971.             60              18
2        47 Product 47          12582.             38              10
3        43 Product 43          10584.             36              10


In [ ]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
supplier_metrics <- suppliers %>%
  select(Supplier_ID, Supplier_Name) %>%
  left_join(
    products %>%
      group_by(Supplier_ID) %>%
      summarise(Product_Count = n_distinct(ProductID), .groups = "drop"),
    by = "Supplier_ID"
  ) %>%
  mutate(Product_Count = coalesce(Product_Count, 0L)) %>%
  select(Supplier_ID, Supplier_Name, Product_Count)
# Required columns: Supplier_ID, Supplier_Name, Product_Count


# Required output for autograding:
cat("Supplier Analysis Summary:
")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "
")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "
")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
   Supplier_ID Supplier_Name Product_Count
7            7    Supplier 7            10
4            4    Supplier 4             6
10          10   Supplier 10             6
1            1    Supplier 1             5
3            3    Supplier 3             5
5            5    Supplier 5             5
6            6    Supplier 6             5
2            2    Supplier 2             3
8            8    Supplier 8             3
9            9    Supplier 9             2


In [ ]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
regional_analysis <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  left_join(customers %>% select(CustomerID, City), by = "CustomerID") %>%
  filter(!is.na(City)) %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(line_total, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    .groups = "drop"
  ) %>%
  mutate(
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)
  ) %>%
  select(City, Total_Sales, Customer_Count, Avg_Customer_Value)
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value


# Required output for autograding:
cat("Regional Analysis Summary:
")
cat("Total cities analyzed:", nrow(regional_analysis), "
")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "
")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             23              3143.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             18              2695.
4 Houston          38143.             18              2119.
5 New York         36417.             16              2276.


In [ ]:
# Your code here:
product_metrics <- order_items %>%
  left_join(products %>% select(ProductID, Product_Name), by = "ProductID") %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  select(ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency) %>%
  arrange(desc(Total_Revenue))

print(head(product_metrics, 10))
# Analyze product performance metrics



# A tibble: 10 × 5
   ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
       <int> <chr>                <dbl>          <int>           <int>
 1        26 Product 26          13971.             60              18
 2        47 Product 47          12582.             38              10
 3        43 Product 43          10584.             36              10
 4        11 Product 11          10186.             36              12
 5        32 Product 32          10105.             35              11
 6        18 Product 18          10019.             25               6
 7        21 Product 21           9725.             30               8
 8        15 Product 15           9674.             32               7
 9        13 Product 13           9470.             31               9
10        17 Product 17           8926.             33               8


In [92]:
# Your code here:

supplier_metrics <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(products %>% select(ProductID, Supplier_ID, Product_Name), by = "ProductID") %>%
  left_join(suppliers %>% select(Supplier_ID, Supplier_Name), by = "Supplier_ID") %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Total_Revenue = sum(line_total, na.rm = TRUE),
    Product_Count = n_distinct(ProductID),
    Order_Count = n_distinct(OrderID),
    Avg_Revenue_per_Product = ifelse(Product_Count > 0, Total_Revenue / Product_Count, NA_real_),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

print(head(supplier_metrics, 10))
# Evaluate supplier performance



# A tibble: 10 × 6
   Supplier_ID Supplier_Name Total_Revenue Product_Count Order_Count
         <int> <chr>                 <dbl>         <int>       <int>
 1           7 Supplier 7           52328.            10          66
 2           5 Supplier 5           50692.             5          54
 3          10 Supplier 10          43450.             6          43
 4           1 Supplier 1           34821.             5          37
 5           3 Supplier 3           30401.             5          36
 6           4 Supplier 4           29898.             6          43
 7           6 Supplier 6           27453.             5          32
 8           9 Supplier 9           17144.             2          16
 9           8 Supplier 8           15497.             3          21
10           2 Supplier 2           11746.             3          20
# ℹ 1 more variable: Avg_Revenue_per_Product <dbl>


In [ ]:
# Your code here:
regional_analysis <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  left_join(customers %>% select(CustomerID, City), by = "CustomerID") %>%
  filter(!is.na(City)) %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(line_total, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    .groups = "drop"
  ) %>%
  mutate(
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)
  ) %>%
  select(City, Total_Sales, Customer_Count, Avg_Customer_Value) %>%
  arrange(desc(Total_Sales))

print(regional_analysis)
# Create regional analysis by customer city



# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             23              3143.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             18              2695.
4 Houston          38143.             18              2119.
5 New York         36417.             16              2276.


In [93]:
# Supplier Performance Analysis
supplier_metrics <- complete_data %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(Supplier_ID, Supplier_Name, Country) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 6
   Supplier_ID Supplier_Name Country     Total_Revenue Products_Supplied
         <int> <chr>         <chr>               <dbl>             <int>
 1           7 Supplier 7    Germany            52328.                10
 2           5 Supplier 5    Japan              50692.                 5
 3          10 Supplier 10   Japan              43450.                 6
 4           1 Supplier 1    USA                34821.                 5
 5           3 Supplier 3    Japan              30401.                 5
 6           4 Supplier 4    South Korea        29898.                 6
 7           6 Supplier 6    Japan              27453.                 5
 8           9 Supplier 9    Japan              17144.                 2
 9           8 Supplier 8    Germany            15497.                 3
10           2 Supplier 2    USA                11746.                 3
# ℹ 1 more variable: Orders_Involved <int>


In [94]:
# Regional Analysis by Customer City
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

Regional Sales Analysis:
# A tibble: 6 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Chicago          72278.             23          36              3143.
2 NA               68753.              5          41             13751.
3 Phoenix          49333.             19          32              2596.
4 Los Angeles      48506.             18          32              2695.
5 Houston          38143.             18          30              2119.
6 New York         36417.             16          26              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [95]:
# REQUIRED variable name: top_customers
# Your code here: Identify top 10% of customers by total spending
# Use quantile() function to find the 90th percentile threshold
library(dplyr)

customer_totals <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  left_join(customers %>% select(CustomerID, Name), by = "CustomerID") %>%
  filter(!is.na(CustomerID)) %>%
  group_by(CustomerID, Name) %>%
  summarise(Total_Spent = sum(line_total, na.rm = TRUE), .groups = "drop")

thresh90 <- quantile(customer_totals$Total_Spent, probs = 0.90, na.rm = TRUE)

top_customers <- customer_totals %>%
  filter(Total_Spent >= thresh90) %>%
  arrange(desc(Total_Spent))

# Required output for autograding (uncomment after completing your analysis):
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)
cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

Customer Segmentation Results:
Top 10% spending threshold: $ 5593.79 
Number of top customers: 10 


In [101]:
# REQUIRED variable name: product_combinations
# Your code here: Find products frequently bought together
# Hint: Group by OrderID, filter orders with multiple items

product_combinations <- order_items %>%
  left_join(products %>% select(ProductID, Product_Name), by = "ProductID") %>%
  group_by(OrderID) %>%
  filter(n() > 1) %>%
  summarise(
    items = list(distinct(tibble(ProductID = ProductID, Product_Name = Product_Name))),
    .groups = "drop"
  ) %>%
  mutate(
    pairs = map(items, ~ {
      it <- .x
      if (nrow(it) < 2) return(tibble())
      cmb <- combn(seq_len(nrow(it)), 2)
      tibble(
        ProductA_ID = it$ProductID[cmb[1,]],
        ProductA_Name = it$Product_Name[cmb[1,]],
        ProductB_ID = it$ProductID[cmb[2,]],
        ProductB_Name = it$Product_Name[cmb[2,]]
      )
    })
  ) %>%
  select(pairs) %>%
  unnest(pairs) %>%
  mutate(
    A_id = pmin(ProductA_ID, ProductB_ID),
    B_id = pmax(ProductA_ID, ProductB_ID),
    A_name = ifelse(ProductA_ID == A_id, ProductA_Name, ProductB_Name),
    B_name = ifelse(ProductB_ID == B_id, ProductB_Name, ProductA_Name)
  ) %>%
  group_by(A_id, A_name, B_id, B_name) %>%
  summarise(Count = n(), .groups = "drop") %>%
  arrange(desc(Count)) %>%
  rename(
    ProductA_ID = A_id,
    ProductA_Name = A_name,
    ProductB_ID = B_id,
    ProductB_Name = B_name
  )

# Required output for autograding (uncomment after completing your analysis):
cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if(nrow(product_combinations) > 0) {
  cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Product Combination Analysis:
Total product combinations found: 281 


Warning message:
"Unknown or uninitialised column: `n`."
Warning message in max(product_combinations$n, na.rm = TRUE):
"no non-missing arguments to max; returning -Inf"


Most frequent combination appears -Inf times


In [104]:
# REQUIRED variable name: critical_suppliers
# Your code here: Analyze which suppliers are most critical
# Consider both revenue contribution and product diversity

temp_suppliers <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(products %>% select(ProductID, Supplier_ID), by = "ProductID") %>%
  left_join(suppliers %>% select(Supplier_ID, Supplier_Name), by = "Supplier_ID") %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Total_Revenue = sum(line_total, na.rm = TRUE),
    Product_Count = n_distinct(ProductID),
    Order_Count = n_distinct(OrderID),
    .groups = "drop"
  )

rev_min <- min(temp_suppliers$Total_Revenue, na.rm = TRUE)
rev_max <- max(temp_suppliers$Total_Revenue, na.rm = TRUE)
prod_min <- min(temp_suppliers$Product_Count, na.rm = TRUE)
prod_max <- max(temp_suppliers$Product_Count, na.rm = TRUE)
total_revenue_all <- sum(temp_suppliers$Total_Revenue, na.rm = TRUE)

critical_suppliers <- temp_suppliers %>%
  mutate(
    Revenue_Share = ifelse(total_revenue_all > 0, Total_Revenue / total_revenue_all, 0),
    Revenue_Score = ifelse(rev_max > rev_min, (Total_Revenue - rev_min) / (rev_max - rev_min), 1),
    Product_Score = ifelse(prod_max > prod_min, (Product_Count - prod_min) / (prod_max - prod_min), 1),
    Composite_Score = 0.7 * Revenue_Score + 0.3 * Product_Score
  ) %>%
  select(Supplier_ID, Supplier_Name, Total_Revenue, Product_Count, Order_Count, Revenue_Share, Composite_Score) %>%
  arrange(desc(Composite_Score))

# Required output for autograding (uncomment after completing your analysis):
cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
print("Top 3 most critical suppliers:")
print(head(critical_suppliers, 3))

Critical Suppliers Analysis:
Total suppliers analyzed: 10 
[1] "Top 3 most critical suppliers:"
# A tibble: 3 × 7
  Supplier_ID Supplier_Name Total_Revenue Product_Count Order_Count
        <int> <chr>                 <dbl>         <int>       <int>
1           1 Supplier 1           34821.             5          37
2           2 Supplier 2           11746.             3          20
3           3 Supplier 3           30401.             5          36
# ℹ 2 more variables: Revenue_Share <dbl>, Composite_Score <dbl>


In [105]:
# REQUIRED variable name: market_expansion  
# Your code here: Identify market expansion opportunities
# Focus on cities with multiple customers but varying order values
market_expansion <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  ) %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  group_by(OrderID, CustomerID) %>%
  summarise(Order_Total = sum(line_total, na.rm = TRUE), .groups = "drop") %>%
  left_join(customers %>% select(CustomerID, City), by = "CustomerID") %>%
  filter(!is.na(City)) %>%
  group_by(City) %>%
  summarise(
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Total_Sales = sum(Order_Total, na.rm = TRUE),
    Avg_Order_Value = ifelse(Order_Count > 0, Total_Sales / Order_Count, 0),
    Order_Value_SD = ifelse(Order_Count > 1, sd(Order_Total, na.rm = TRUE), 0),
    Order_Value_CV = ifelse(Avg_Order_Value > 0, Order_Value_SD / Avg_Order_Value, 0),
    Avg_Spent_per_Customer = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0),
    .groups = "drop"
  ) %>%
  filter(Customer_Count >= 2) %>%            # cities with multiple customers
  arrange(desc(Order_Value_CV))              # high variation first

# Required output for autograding (uncomment after completing your analysis):
cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))

Market Expansion Analysis:
Cities evaluated for expansion: 5 
[1] "Top expansion opportunities:"
# A tibble: 5 × 8
  City     Customer_Count Order_Count Total_Sales Avg_Order_Value Order_Value_SD
  <chr>             <int>       <int>       <dbl>           <dbl>          <dbl>
1 New York             16          26      36417.           1401.          1125.
2 Phoenix              19          32      49333.           1542.          1177.
3 Los Ang…             18          32      48506.           1516.          1153.
4 Chicago              23          36      72278.           2008.          1458.
5 Houston              18          30      38143.           1271.           853.
# ℹ 2 more variables: Order_Value_CV <dbl>, Avg_Spent_per_Customer <dbl>


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** [Report specific percentages of matched records]
- **Data Relationships:** [Describe the relationships you discovered between tables]
- **Multi-table Joins:** [Explain how the data grew/changed through the join sequence]

### Data Quality Issues Discovered:
- **Orphaned Records:** [Report specific counts of unmatched records]
- **Missing Data:** [Identify which tables had missing or invalid references]
- **Referential Integrity:** [Describe any foreign key violations found]

### Business Insights:
- **Top Customer:** [Name and spending amount of highest value customer]
- **Best Product:** [Product name and revenue from your analysis]
- **Regional Performance:** [Key findings about city-level sales patterns]
- **Supplier Analysis:** [Insights about supplier diversity and concentration]

### Strategic Recommendations:
- [Provide at least 3 specific, actionable recommendations based on your data analysis]

### Technical Learnings:
- **Inner vs Left Joins:** [When you used each and why]
- **Multi-table Strategy:** [Your approach to building complex joined datasets]
- **Performance Considerations:** [Observations about join efficiency and data size]

## Analysis Report

The join analysis shows high match rates for transactional lines but gaps in referential keys: 100% of order_items (400 rows) matched to orders, while 80.0% of orders (200 of 250) matched to a customer (leaving 50 orphaned orders). The tables exhibit expected one-to-many relationships (customers → orders → order_items) and many-to-one lookups for products → suppliers; joining order_items to orders preserved the 400 line-level records and adding customer details resolved customer-level attributes for all matched orders. Data quality issues include 50 orphaned orders (orders referencing CustomerID values not present in customers) and no unordered products (0 products never ordered), indicating missing or invalid customer references as the primary integrity problem. Business insights: the highest-value customer is Customer 103 with total spending of $17,221.69; the top-selling product by revenue is Product 26 with $13,970.66 in sales; regionally, Chicago leads with $72,277.52 in total sales; supplier concentration is material — the top three suppliers (Supplier 7, Supplier 5, Supplier 10) account for 46.73% of total supplier-attributed revenue. Strategic recommendations: (1) investigate and correct the 50 orphaned orders (verify CustomerID capture, run a reconciliation with source systems) to restore full customer-level analytics and avoid lost follow-up opportunities; (2) prioritize inventory and marketing for Product 26 and the other top revenue SKUs while testing price/bundle experiments on high-volume, lower-revenue items to improve margin; (3) mitigate supplier concentration risk by negotiating contingency supply agreements or qualifying alternative suppliers for the top 3 suppliers that together supply nearly half of revenue. Technical learnings: use left_joins when preserving line-level sales is required (order_items left_join orders/customers) and inner_joins when you explicitly want to restrict analysis to fully matched records; compute global summary scalars (min/max/total) outside rowwise mutate calls for stable scoring; and for performance at scale ensure join keys are indexed, perform joins in order from highest-cardinality (lines) to lower-cardinality (lookups), and prefer vectorized aggregations rather than rowwise operations to keep multi-table joins efficient.

In [107]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired
# Ensure numeric types and compute line_total
order_items <- order_items %>%
  mutate(
    Quantity = as.numeric(Quantity),
    Unit_Price = as.numeric(Unit_Price),
    line_total = coalesce(Quantity, 0) * coalesce(Unit_Price, 0)
  )

orders <- orders %>%
  mutate(
    OrderID = as.integer(OrderID),
    CustomerID = as.integer(CustomerID)
  )

customers <- customers %>%
  mutate(CustomerID = as.integer(CustomerID))

products <- products %>%
  mutate(ProductID = as.integer(ProductID),
         Supplier_ID = as.integer(Supplier_ID))

suppliers <- suppliers %>%
  mutate(Supplier_ID = as.integer(Supplier_ID))

# Basic counts and match rates
total_order_items <- nrow(order_items)
order_items_matched_orders <- sum(order_items$OrderID %in% orders$OrderID)
order_items_match_pct <- 100 * order_items_matched_orders / total_order_items

total_orders <- nrow(orders)
orders_matched_customers <- sum(orders$CustomerID %in% customers$CustomerID)
orders_match_pct <- 100 * orders_matched_customers / total_orders

total_customers <- nrow(customers)
customers_with_orders <- sum(customers$CustomerID %in% orders$CustomerID)
customers_no_orders <- total_customers - customers_with_orders

orphaned_orders_count <- total_orders - orders_matched_customers
orphaned_orders_ids <- orders$OrderID[!orders$CustomerID %in% customers$CustomerID]

products_never_ordered_count <- sum(!products$ProductID %in% order_items$ProductID)
products_never_ordered_ids <- products$ProductID[!products$ProductID %in% order_items$ProductID]

# Top customer by total spending
cust_spend <- order_items %>%
  inner_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  group_by(CustomerID) %>%
  summarise(Total_Spent = sum(line_total, na.rm = TRUE), .groups = "drop") %>%
  left_join(customers %>% select(CustomerID, Name), by = "CustomerID") %>%
  arrange(desc(Total_Spent))

top_customer_name <- cust_spend$Name[1]
top_customer_spend <- cust_spend$Total_Spent[1]

# Best product by revenue
product_revenue <- order_items %>%
  group_by(ProductID) %>%
  summarise(Total_Revenue = sum(line_total, na.rm = TRUE), Total_Quantity = sum(Quantity, na.rm = TRUE), Order_Frequency = n_distinct(OrderID), .groups = "drop") %>%
  left_join(products %>% select(ProductID, Product_Name), by = "ProductID") %>%
  arrange(desc(Total_Revenue))

best_product_name <- product_revenue$Product_Name[1]
best_product_rev <- product_revenue$Total_Revenue[1]

# Regional (city) sales patterns
order_totals <- order_items %>%
  group_by(OrderID) %>%
  summarise(Order_Total = sum(line_total, na.rm = TRUE), .groups = "drop")

order_city <- order_totals %>%
  left_join(orders %>% select(OrderID, CustomerID), by = "OrderID") %>%
  left_join(customers %>% select(CustomerID, City), by = "CustomerID")

city_stats <- order_city %>%
  filter(!is.na(City)) %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Order_Total, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    Customer_Count = n_distinct(CustomerID),
    Avg_Order_Value = ifelse(Order_Count > 0, Total_Sales / Order_Count, 0),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

top_city <- city_stats$City[1]
top_city_total_sales <- city_stats$Total_Sales[1]

# Supplier analysis: revenue and diversity
supplier_stats <- order_items %>%
  left_join(products %>% select(ProductID, Supplier_ID), by = "ProductID") %>%
  group_by(Supplier_ID) %>%
  summarise(
    Total_Revenue = sum(line_total, na.rm = TRUE),
    Product_Count = n_distinct(ProductID),
    Order_Count = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  left_join(suppliers %>% select(Supplier_ID, Supplier_Name), by = "Supplier_ID") %>%
  arrange(desc(Total_Revenue))

total_revenue_all <- sum(supplier_stats$Total_Revenue, na.rm = TRUE)
top3_suppliers <- supplier_stats %>% slice_head(n = 3)
top3_rev_share_pct <- if (total_revenue_all > 0) 100 * sum(top3_suppliers$Total_Revenue, na.rm = TRUE) / total_revenue_all else 0

# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")


=== HOMEWORK 6 COMPLETION SUMMARY ===
✓ Data Import: 5 datasets loaded
✓ Basic Joins: 4 join types completed
✓ Multi-table Joins: 4-step progression completed
✓ Data Quality: Anti-joins and semi-joins performed
✓ Business Analysis: Customer, product, supplier, and regional analysis completed
✓ Complex Questions: Advanced business scenarios analyzed
✓ Summary: Comprehensive insights documented

All required variables created for autograding verification.
